### PR TITLE
Strategic AI economy update

### PR DESCRIPTION
**Pull request by idshibanov**
Adding a kingdom budget into consideration. AI heroes now should focus on collecting missing resources based on current castle build order.

### CODE REVIEW

File: src/fheroes2/ai/normal/ai_normal_castle.cpp [hunk 1]
```diff
@@ -192,10 +193,53 @@ namespace AI
...
+            if ( castle->isBuild( DWELLING_MONSTER6 ) ) {
+                Funds bestUnitCost = Monster( race, DWELLING_MONSTER6 ).GetUpgrade().GetCost();
+                for ( BudgetEntry & budgetEntry : _budget ) {
+                    if ( bestUnitCost.Get( budgetEntry.resource ) > 0 ) {
+                        budgetEntry.recurringCost = true;
+                    }
+                }
+            }
+        }
+        for ( BudgetEntry & budgetEntry : _budget ) {
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_castle.cpp:230]: Please add a separate line to differentiate 2 loops.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @idshibanov , I put just few comments here. Could you please check them?

**Comment by ihhub** [comment]
Hi @idshibanov , could you please address given comment in this pull request once you have spare time? We would like to merge these changes for the upcoming release.

**Comment by idshibanov** [comment]
> Hi @idshibanov , could you please address given comment in this pull request once you have spare time? We would like to merge these changes for the upcoming release.

I've delayed coming back to this change because I wasn't totally happy with the AI's behaviour after. The values might need some fine tuning. I'll find some time.

File: src/fheroes2/ai/normal/ai_normal_castle.cpp [hunk 2]
```diff
@@ -192,10 +193,52 @@ namespace AI
...
+            }
+
+            if ( castle->isBuild( DWELLING_MONSTER6 ) ) {
+                Funds bestUnitCost = Monster( race, DWELLING_MONSTER6 ).GetUpgrade().GetCost();
+                for ( BudgetEntry & budgetEntry : _budget ) {
+                    if ( bestUnitCost.Get( budgetEntry.resource ) > 0 ) {
+                        budgetEntry.recurringCost = true;
+                    }
+                }
+            }
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_castle.cpp:226]: I think this condition needs an adjustment that not all castles allow to build monster dwellings of level 6.

File: src/fheroes2/ai/normal/ai_normal_castle.cpp [hunk 3]
```diff
@@ -192,10 +193,52 @@ namespace AI
...
+            if ( !castle ) {
+                continue;
+            }
+
+            const int race = castle->GetRace();
+            const std::vector<BuildOrder> & buildOrder = GetBuildOrder( race );
+            for ( const BuildOrder & order : buildOrder ) {
+                const int status = castle->CheckBuyBuilding( order.building );
+                if ( status == LACK_RESOURCES ) {
+                    requirements = requirements.max( PaymentConditions::BuyBuilding( race, order.building ) );
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_castle.cpp:215]: We use payment conditions for a building as a requirement. Are we taking into account that for instance only one resource might be required like Gems and another case that we need only small part of the building price. For example, we need only extra 100 gold.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @idshibanov , I left 2 questions in this pull request. Could you please check them when you have spare time?

File: src/fheroes2/ai/normal/ai_normal_castle.cpp (hunk 2 above)
Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_castle.cpp:228]: Thanks for the explanation!

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by ihhub** [comment]
@idshibanov , thank you so much for this improvement!

**Pull request merged**